src/Oscoin/Protocol/Sync.hs

```diff
@@ -0,0 +1,393 @@
+    , scDataFetcher       :: DataFetcher c tx s m
+    -- ^ A 'DataFetcher'.
+    , scEventTracer       :: Tracer m
+    -- ^ A tracer for interesting events that happens during syncing.
+    , scConcurrently      :: forall a b t.  Traversable t => t a -> (a -> m b) -> m (t b)
+    -- ^ An action that can be used to fetch some resources in parallel.
+    -- 'IO'-based implementations will probably use 'forConcurrently' or some
+    -- variation of it.
+    , scLocalChainReader  :: BlockStoreReader c tx s m
+    -- ^ A read-only handle to the local blockchain.
+    , scUpstreamConsumers :: [SyncEvent c tx s -> m ()]
+    -- ^ A list of \"upstream consumers\", represented as effectful actions from
+    -- a 'SyncEvent' to 'm ()'.
+    }
+
+-- | A sync monad over @m@, that can throw 'SyncError's.
+newtype Sync c tx s m a =
+    Sync { runSync :: ExceptT SyncError (ReaderT (SyncContext c tx s m) m) a }
+    deriving ( Functor
+             , Applicative
+             , Monad
+             , MonadError SyncError
+             , MonadReader (SyncContext c tx s m)
+             )
+
+instance MonadTrans (Sync c tx s) where
+    lift = Sync . lift . lift
+
+instance MonadIO m => MonadIO (Sync c tx s m) where
+    liftIO = Sync . liftIO
+
+{------------------------------------------------------------------------------
+  Convenience functions over Sync
+------------------------------------------------------------------------------}
+
+withActivePeers
+    :: Monad m
+    => (ActivePeers c -> Sync c tx s m a)
+    -> Sync c tx s m a
+withActivePeers f = do
+    ctx <- ask
+    activePeers <- lift (scActivePeers ctx)
+    when (HS.null activePeers) $ throwError NoActivePeers
+    f activePeers
+
+-- | Like 'withActivePeers', but it passes to the callback only a (single)
+-- active peer. This might be useful in tests which it's not necessary to have
+-- all the active peers available.
+withActivePeer
+    :: Monad m
+    => (ActivePeer c -> Sync c tx s m a)
+    -> Sync c tx s m a
+withActivePeer f = do
+    ctx <- ask
+    activePeers <- lift (scActivePeers ctx)
+    case HS.toList activePeers of
+      []           -> throwError NoActivePeers
+      activePeer:_ -> f activePeer
+
+{------------------------------------------------------------------------------
+  Pure functions & constants
+------------------------------------------------------------------------------}
+
+height :: Block c tx s -> Height
+height = blockHeight . blockHeader
+
+-- | The nu parameter, from the spec.
+type Nu = Natural
+
+maxBlocksPerPeer :: Int
+maxBlocksPerPeer = 50
+
+-- | Returns 'True' if the node finished syncing all the blocks.
+-- This is the 'isDone' state transition function from the spec.
+isDone
+    :: ( Eq tx
+       , Eq s
+       , Eq (Hash c)
+       , Eq (PublicKey c)
+       )
+    => Nu
+    -> Block c tx s
+    -> Block c tx s
+    -> Bool
+isDone (fromIntegral -> nu) localTip remoteTip =
+    remoteTip == localTip || height remoteTip - height localTip < nu
```

Comment:
Might be a good idea to handle the case where `localTip` is ahead of `remoteTip`, or mark an invariant. Even if `Height` is currently an `Integer` and can go in the negative, if that ever changes, this could silently break.

Reply:
I think you're right -- not just for the silent breakage, but also because if we are ahead, it might mean there was a fork, so we _do want_ to sync again (i.e. `isDone` should yield `False` in that case).